### DISCUSSION TITLE
routerAdd is 404, middleware triggers just fine

### DISCUSSION THREAD

**Discussion by a4amado** [pocketbase/pocketbase#6112] [Q&A] [answered]
```json
{"data":{},"message":"The requested resource wasn't found.","status":404}
```

```
Request URL:http://127.0.0.1:8090/api/exchange/start_chat/h0kl3pw928yxfk8/
Request Method: POST
Status Code: 404 Not Found
Remote Address: 127.0.0.1:8090
Referrer Policy: strict-origin-when-cross-origin
```

```js

routerAdd(
  "POST",
  "/api/exchange/start_chat/{user_id}",
  async (c) => {
    return c.json(200, { message: "Chat started successfully!" });
  },
);

```

```tsx

    <button
      onClick={() => {
        pb.collection("users")
          .authWithPassword("a4asas@sd.com", "123456789")
          .then(() => {
            const token = JSON.parse(localStorage.getItem("pocketbase_auth") || "").token
            
            axios({                           
              url: `http://127.0.0.1:8090/api/exchange/start_chat/h0kl3pw928yxfk8/`,
              headers: {
                Authorization:token
              },
              method: "POST"
            }).then((e) => {
              console.log(e);
            });
          });
      }}
    >

      sss
    </button>
```

**Comment by a4amado** [comment, marked as answer]
LOOOOOOL, it's was the `trailing slash` in the URL

**Comment by muhammadabdulhamidkpanje** [comment]
I am facing the same issue here
routerAdd("POST", "data/export/", async (c) => {
  try {
    const record = c.get("authRecord");
    const xlsx = require("xlsx/dist/xlsx.full.min.js");

    if (!record) {
      throw new UnauthorizedError("You must be logged in to export data");
    }

    // if (!["national", "local"].includes(record.role)) {
    //   throw new UnauthorizedError("You do not have permission to export data");
    // }

    const utils = require(`${__hooks}/utils.js`);

    // Validate the `template` parameter
    const template = c.request.url.query().get("template");
    if (!template || !/^[a-zA-Z_]+$/.test(template)) {
      throw new Error("Invalid template specified");
    }

    // Query the database
    const result = [];
    const query = `SELECT * FROM ${template} LIMIT 100`;
    await $app.db()
      .newQuery(query)
      .all(result);

    if (result.length === 0) {
      return c.json(404, {
        message: "No data found for the specified template",
        success: false,
      });
    }

    // Create Excel file
    const worksheet = xlsx.utils.json_to_sheet(result);
    const workbook = xlsx.utils.book_new();
    xlsx.utils.book_append_sheet(workbook, worksheet, template);

    const exportBuffer = xlsx.write(workbook, { type: "buffer", bookType: "xlsx" });

    // Send the Excel file as a response
    c.response.type = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
    c.response.headers.set(
      "Content-Disposition",
      `attachment; filename="${template}.xlsx"`
    );
    return c.send(201, exportBuffer);

  } catch (error) {
    console.error("Export Error:", error);

    return c.json(400, {
      message: error.message || "An unexpected error occurred",
      success: false,
      details: error.stack || "No additional details",
    });
  }
}, $apis.requireAdminOrRecordAuth());
routerAdd("POST", "data/export/", async (c) => {
  try {
    const record = c.get("authRecord");
    const xlsx = require("xlsx/dist/xlsx.full.min.js");

    if (!record) {
      throw new UnauthorizedError("You must be logged in to export data");
    }

    // if (!["national", "local"].includes(record.role)) {
    //   throw new UnauthorizedError("You do not have permission to export data");
    // }

    const utils = require(`${__hooks}/utils.js`);

    // Validate the `template` parameter
    const template = c.request.url.query().get("template");
    if (!template || !/^[a-zA-Z_]+$/.test(template)) {
      throw new Error("Invalid template specified");
    }

    // Query the database
    const result = [];
    const query = `SELECT * FROM ${template} LIMIT 100`;
    await $app.db()
      .newQuery(query)
      .all(result);

    if (result.length === 0) {
      return c.json(404, {
        message: "No data found for the specified template",
        success: false,
      });
    }

    // Create Excel file
    const worksheet = xlsx.utils.json_to_sheet(result);
    const workbook = xlsx.utils.book_new();
    xlsx.utils.book_append_sheet(workbook, worksheet, template);

    const exportBuffer = xlsx.write(workbook, { type: "buffer", bookType: "xlsx" });

    // Send the Excel file as a response
    c.response.type = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
    c.response.headers.set(
      "Content-Disposition",
      `attachment; filename="${template}.xlsx"`
    );
    return c.send(201, exportBuffer);

  } catch (error) {
    console.error("Export Error:", error);

    return c.json(400, {
      message: error.message || "An unexpected error occurred",
      success: false,
      details: error.stack || "No additional details",
    });
  }
}, $apis.requireAdminOrRecordAuth());




import {
  Box,
  Button,
  Flex,
  Heading,
  Icon,
  Select,
  Text,
  useToast,
} from "@chakra-ui/react";
import { TemplateAttr, ExportTemplates as Templates } from "../data-upload/templates";
import { useState } from "react";
import { MdDownload } from "react-icons/md";
import { pocketbase } from "../../pocketbase";

export default function DataExport() {
  const [selectedTemplate, setSelectedTemplate] = useState<
    Partial<TemplateAttr>
  >({});
  const [isExporting, setIsExporting] = useState(false);
  const toast = useToast();

  const handleExportData = async () => {
    if (!selectedTemplate.name || isExporting) return;

    setIsExporting(true);

    try {

      const formData = new FormData();
      formData.append("text", selectedTemplate.name )
      const token = pocketbase.authStore.token;
      const response = await pocketbase.send(`/data/export?template=${selectedTemplate.name}`,
        {
          method: "GET",
          headers: {
            Authorization: `Bearer ${token}`,
          },
        }
      );

      if (!response.ok) {
        throw new Error("Failed to export data.");
      }

      const blob = await response.blob();
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement("a");
      link.href = url;
      link.setAttribute(
        "download",
        `${selectedTemplate.name.replace(/\s+/g, "_")}.csv`
      );
      document.body.appendChild(link);
      link.click();
      link.parentNode?.removeChild(link);

      toast({
        title: "Export Successful",
        description: `Data for "${selectedTemplate.name}" exported successfully.`,
        status: "success",
        duration: 2000,
        isClosable: true,
        position: "bottom",
      });
    } catch (error) {
      console.error("Export error:", error);
      toast({
        title: "Error",
        description: "Failed to export data.",
        status: "error",
        duration: 2000,
        isClosable: true,
        position: "bottom",
      });
    } finally {
      setIsExporting(false);
    }
  };

  return (
    <Box flex={1}>
      <Box
        p={4}
        maxW="80%"
        mx="auto"
        h="100%"
        bg="gray.50"
        borderRadius="md"
        boxShadow="md"
      >
        <Heading mb={4} fontSize="2xl" fontWeight="bold" color="secondary">
          Data Export
        </Heading>
        <Text mb={4} fontSize="lg" color="gray.700">
          Export data from the system here:
        </Text>

        <Flex flexDirection="column" p={4} mx={2} my={5} columnGap={3}>
          <Select
            placeholder="Select template"
            onChange={(e) => {
              const found = Templates.find(
                (template) => template.id === e.target.value
              );
              if (found) {
                setSelectedTemplate(found);
              }
            }}
            pb = "8px"
            border="2px"
            borderColor="secondary"
            borderRadius="full"
            boxShadow="sm"
            fontSize="lg"
            _hover={{ borderColor: "teal.500" }}
            _focus={{ borderColor: "teal.700", boxShadow: "outline" }}
          >
            {Templates.map((template) => (
              <option key={template.id} value={template.id}>
                {template.name}
              </option>
            ))}
          </Select>

          <Button
            disabled={!selectedTemplate.name}
            bg="secondary"
            isLoading={isExporting}
            color="white"
            p={4}
            px={8}
            borderRadius="md"
            boxShadow="sm"
            onClick={handleExportData}
            _hover={{ bg: "teal.600" }}
            _active={{ bg: "teal.700" }}
          >
            <Icon as={MdDownload} w="20px" h="20px" mr={2} />
            Export
          </Button>
        </Flex>
      </Box>
    </Box>
  );
}
